src/zarr/v3/store/memory.py

```diff
@@ -67,20 +68,16 @@ async def delete(self, key: str) -> None:
     async def set_partial_values(self, key_start_values: List[Tuple[str, int, bytes]]) -> None:
         raise NotImplementedError
 
-    async def list(self) -> List[str]:
-        return list(self._store_dict.keys())
+    async def list(self) -> AsyncGenerator[str, None]:
+        for key in self._store_dict:
+            yield key
 
-    async def list_prefix(self, prefix: str) -> List[str]:
-        return [key for key in self._store_dict if key.startswith(prefix)]
+    async def list_prefix(self, prefix: str) -> AsyncGenerator[str, None]:
+        for key in self._store_dict:
+            if key.startswith(prefix):
+                yield key
 
-    async def list_dir(self, prefix: str) -> List[str]:
-        if prefix == "":
-            return list({key.split("/", maxsplit=1)[0] for key in self._store_dict})
-        else:
-            return list(
-                {
-                    key.removeprefix(prefix + "/").split("/")[0]
-                    for key in self._store_dict
-                    if (key.startswith(prefix + "/") and key != prefix)
-                }
-            )
+    async def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
+        for key in self._store_dict:
+            if key.startswith(prefix + "/") and key != prefix:
+                yield key.strip(prefix + "/").rsplit("/", maxsplit=1)[0]
```

Comment:
Confusingly, `key.strip("foo")` treats "foo" as a set, e.g. `{"f", "o"}`: 

```python
>>> 'ba/x/bbaa'.strip('ab')
'/x/'
``` 

I think `removeprefix` does what we want here:

```python
>>> 'ba/x/bbaa'.removeprefix('ab')
'ba/x/bbaa'
>>> 'ba/x/bbaa'.removeprefix('ba')
'/x/bbaa'
```

```suggestion
                yield key.removeprefix(prefix + "/").rsplit("/", maxsplit=1)[0]
```